Implement Execution Comparison

vladarama opened the pull request:
This PR is a cleanup based on @farajidaneshgar work on the execution comparison ([Original PR]).
The cleanup includes fixing `Eclipse` and `SonarLint` warnings and includes some general code quality and formatting improvements.

Signed-off-by: Vlad Arama <vlad.arama@ericsson.com>

[Original PR]: https://github.com/eclipse-tracecompass-incubator/org.eclipse.tracecompass.incubator/pull/13

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+    protected static final Logger LOGGER = Logger.getLogger(DifferentialFlameGraphView.class.getName());
+
+    private @Nullable DifferentialWeightedTreeProvider<?> fDataProvider;
+    private @Nullable ITimeGraphDataProvider<TimeGraphEntryModel> fdataProviderGroup = null;
```
Should be FlameChartEntryModel instead of TimeGraphEntryModel

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+                    return;
+                }
+                fDataProvider = dataProvider;
+                setFdataProviderGroup(new FlameGraphDataProvider(trace, fDataProvider, FlameGraphDataProvider.ID + ':' + DifferentialSeqCallGraphAnalysis.ID));
```
Replace with `new FlameGraphDataProvider<>(...`

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+                while (!complete && !monitor.isCanceled()) {
+                    Map<String, Object> parameters = new HashMap<>(additionalParams);
+                    parameters.put(DataProviderParameterUtils.REQUESTED_TIME_KEY, ImmutableList.of(0, Long.MAX_VALUE));
+                    TmfModelResponse<TmfTreeModel<TimeGraphEntryModel>> responseGroupA = getFdataProviderGroup().fetchTree(parameters, monitor);
```
Replace TimeGraphEntryModel with FlameChartEntryModel.

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+                    }
+
+                    complete = responseGroupA.getStatus() == ITmfResponse.Status.COMPLETED;
+                    TmfTreeModel<TimeGraphEntryModel> groupAModel = responseGroupA.getModel();
```
Replace TimeGraphEntryModel with FlameChartEntryModel.

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+            return displayWidth <= 0 ? 1 : displayWidth;
+        }
+
+        private ITimeGraphDataProvider<TimeGraphEntryModel> getFdataProviderGroup() {
```
Replace TimeGraphEntryModel with FlameChartEntryModel.

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+            return fdataProviderGroup;
+        }
+
+        private void setFdataProviderGroup(@Nullable ITimeGraphDataProvider<TimeGraphEntryModel> dataProviderGroup) {
```
Replace TimeGraphEntryModel with FlameChartEntryModel.

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+    /**
+     * Zoom thread
+     *
+     * @since 1.1
```
Remove since comment. Improve comment text to describe what this thread is used for.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+     * @param entry
+     *            queried {@link TimeGraphEntry}.
+     * @return the {@link ITimeGraphDataProvider}
+     * @since 3.3
```
Remove since comment

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+    }
+
+    private void redraw() {
+
```
remove space

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+    }
+
+    private Set<TimeGraphEntry> getVisibleItems(int buffer) {
+
```
remove space

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+
+    private static BiFunction<ITimeEvent, Long, Map<String, String>> getTooltipResolver(ITimeGraphDataProvider<? extends TimeGraphEntryModel> provider) {
+        return (event, time) -> getTooltip(event, time, provider, false);
+
```
remove space

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+            }
+        }
+
+        private void buildEntryList(@Nullable ITmfTrace trace, ITmfTrace parentTrace, Map<String, Object> additionalParams, IProgressMonitor monitor) {
```
Refactor this method to reduce complexity, lines of code and nesting level.

vladarama replied:
Done, its much better than the initial implementation, but we are still at a Cognitive Complexity level of 18. 
I am struggling to refactor it more than that, what do you think ?

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+                .collect(Collectors.toSet());
+    }
+
+    private class BuildRunnable {
```
Refactor this private class in another file to reduce the complexity of the DifferentialFlameGraphView.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+
+                }
+            } catch (InterruptedException e1) {
+                // TODO Auto-generated catch block
```
Replace with log call from Activator (e.g. Activator.getDefault().logError())

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+    }
+
+    private void zoomEntries(Map<Long, TimeGraphEntry> map, List<ITimeGraphRowModel> model, boolean completed, Sampling sampling) {
+        boolean isZoomThread = false;
```
Remove

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+
+            if (entry != null) {
+                List<ITimeEvent> events = createTimeEvents(entry, rowModel.getStates());
+                if (isZoomThread) {
```
remove

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1261 @@
+                    }
+                } else {
+
+                    entry.setEventList(events);
```
remove

arfio reviewed (changes requested):
I just wrote a few comments on DifferentialFlameGraphView, I will do another pass tomorrow. If you have any questions/remarks we can discuss that in the comment threads.

vladarama commented on callstack/org.eclipse.tracecompass.incubator.callstack.core/src/org/eclipse/tracecompass/incubator/internal/callstack/core/instrumented/callgraph/AggregatedCalledFunction.java:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,23 @@
+ org.eclipse.tracecompass.analysis.profiling.core,
+ org.eclipse.jdt.annotation;bundle-version="[2.0.0,3.0.0)";resolution:=optional
+Export-Package: org.eclipse.tracecompass.incubator.internal.executioncomparison.ui;x-internal:=true
+Automatic-Module-Name: org.eclipse.tracecompass.incubator.executioncomparision.ui
```
typo

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/build.properties:
```diff
@@ -0,0 +1,18 @@
+###############################################################################
+# Copyright (c) 2022 École Polytechnique de Montréal
```
2024

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/plugin.properties:
```diff
@@ -0,0 +1,14 @@
+###############################################################################
+# Copyright (c) 2022 École Polytechnique de Montréal
```
2024

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/plugin.properties:
```diff
@@ -0,0 +1,14 @@
+###############################################################################
+
+Bundle-Vendor = Eclipse Trace Compass Incubator
+Bundle-Name = Trace Compass ExecutionComparision UI Plug-in (Incubator)
```
typo

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision/build.properties:
```diff
@@ -0,0 +1,13 @@
+###############################################################################
+# Copyright (c) 2022 École Polytechnique de Montréal
```
2024

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision/feature.properties:
```diff
@@ -0,0 +1,18 @@
+###############################################################################
+# Copyright (c) 2022 École Polytechnique de Montréal
```
2024

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision/feature.properties:
```diff
@@ -0,0 +1,18 @@
+# SPDX-License-Identifier: EPL-2.0
+###############################################################################
+
+featureName=Trace Compass ExecutionComparision (Incubation)
```
typo

arfio commented on callstack/org.eclipse.tracecompass.incubator.callstack.core/META-INF/MANIFEST.MF:
```diff
@@ -44,5 +44,6 @@ Import-Package: com.google.common.annotations,
  com.google.common.collect,
- org.apache.commons.lang3
+ org.apache.commons.lang3,
+ org.eclipse.tracecompass.incubator.internal.callstack.core.instrumented.callgraph
```
This change still needs to be removed.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+
+    @Override
+    protected boolean executeAnalysis(IProgressMonitor monitor) throws TmfAnalysisException {
+        return true;
```
Should be implemented correctly: run the analysis and return true if analysis completed successfully, false otherwise.

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+
+    @Override
+    public boolean waitForCompletion() {
+        return true;
```
Should be implemented correctly look at other analyses implementation to see how it is typically done.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+
+    @Override
+    protected void canceling() {
+        // Does nothing
```
Should cancel the analysis and stop any running thread. (through the use of a monitor for example, refer to other analysis implementation)

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+
+    @Override
+    public boolean canExecute(ITmfTrace trace) {
+        return (trace instanceof TmfExperiment);
```
Should look at the traces included in the experiment and see if they provide any analysis that are in the fCallStackAnalysisMap.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+
+    private static final String MERGE = "Merge"; //$NON-NLS-1$
+    private @Nullable DifferentialCallGraphProvider fDifferentialCallGraphProvider;
+    private static Map<String, String> fcallStackAnalysisMap = new HashMap<>();
```
Should be fCallStackAnalysisMap

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+    public static final String ID = "org.eclipse.tracecompass.incubator.executioncomparison.diffcallgraph"; //$NON-NLS-1$
+    private static final Logger LOGGER = TraceCompassLog.getLogger(DifferentialSeqCallGraphAnalysis.class);
+
+    private static final String MERGE = "Merge"; //$NON-NLS-1$
```
The static variables should be grouped together and the other attributes should be grouped separately.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+     * @param monitor
+     *            the progress monitor
+     */
+    public void refreshDiffCG(@Nullable IProgressMonitor monitor) {
```
The monitor should not be @Nullable

arfio replied:
This method should return the fDifferentialCallGraphProvider.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+     *
+     * @return DifferentialCallGraph
+     */
+    public IWeightedTreeSet<ICallStackSymbol, Object, DifferentialWeightedTree<ICallStackSymbol>> getCallGraph() {
```
Should be called getDifferentialCallGraph().

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+     * @return DifferentialCallGraph
+     */
+    public IWeightedTreeSet<ICallStackSymbol, Object, DifferentialWeightedTree<ICallStackSymbol>> getCallGraph() {
+        DifferentialCallGraphProvider differentialCallGraphProvider = getDifferentialCallGraphProvider();
```
Should schedule the analysis if the provider is null.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+     *            the progress monitor, can be null
+     * @return the differential weighted provider or null
+     */
+    public @Nullable DifferentialWeightedTreeProvider<?> getDiffProvider(@Nullable IProgressMonitor monitor) {
```
Should be called getDifferentialTreeProvider or other that explicitly tells why it is different from getCallGraph() ?

arfio replied:
Monitor should not be nullable.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+                    if (fDiffJob != null) {
+                        fDiffJob.cancel();
+                    }
+                    fDiffJob = new Job("Make differential Callgraph") { //$NON-NLS-1$
```
This job should be part of the executeAnalysis mechanism and other inherited methods from TmfAbstractAnalysisModule.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+    }
+
+    @Override
+    public void dispose() {
```
This method should cancel all running jobs that are running outside of the executeAnalysis method or that call directly executeAnalysis(). See the implementation of super.dispose() for more details.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+     *
+     * @return the differentialCallGraphProvider
+     */
+    private @Nullable DifferentialCallGraphProvider getDifferentialCallGraphProvider() {
```
Delete this method and replace with direct access to fDifferentialCallGraphProvider

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+     * @param differentialCallGraphProvider
+     *            the differentialCallGraphProvider to set
+     */
+    private void setDifferentialCallGraphProvider(DifferentialCallGraphProvider differentialCallGraphProvider) {
```
Delete this method and replace with direct access to fDifferentialCallGraphProvider

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+                        callGraphModule = module;
+                        callGraphModule.schedule();
+                        fTraceCallGraphRegistry.put(String.valueOf(traceMember.getName()), callGraphModule);
+                        callGraphModule.waitForCompletion();
```
This loop should look at monitor and exits and cancel any jobs if the monitor is cancelled.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/DifferentialSeqCallGraphAnalysis.java:
```diff
@@ -0,0 +1,409 @@
+     *            the progress monitor, can be null
+     * @return the differential weighted provider or null
+     */
+    public @Nullable DifferentialWeightedTreeProvider<?> getDiffProvider(@Nullable IProgressMonitor monitor) {
```
Should not be nullable ?

vladarama replied:
This method returns `fDifferentialCallGraphProvider` which is nullable. We could change `fDifferentialCallGraphProvider` to be Non-Nullable, but it would require changing the implementation a bit. I think the current implementation works fine, but we could definitely change it if you think its better.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/ExecutionComparisonView.java:
```diff
@@ -0,0 +1,1212 @@
+    /**
+     * The title of the view
+     */
+    @SuppressWarnings("null")
```
remove and use Objects.requireNonNull instead.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/ExecutionComparisonView.java:
```diff
@@ -0,0 +1,1212 @@
+    private ITmfTimestamp fEndTimeA = TmfTimestamp.BIG_CRUNCH;
+    private ITmfTimestamp fEndTimeB = TmfTimestamp.BIG_CRUNCH;
+    @SuppressWarnings("null")
+    private String fStatistic = Messages.multipleDensityViewDuration;
```
Should be static final and use Objects.requireNonNull

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/ExecutionComparisonView.java:
```diff
@@ -0,0 +1,1212 @@
+    private @Nullable TmfViewer fTmfViewerB;
+    private List<String> fTraceListA = new ArrayList<>();
+    private List<String> fTraceListB = new ArrayList<>();
+    private String fTraceStr = "Trace(s): ";//$NON-NLS-1$
```
Should be static final and use Messages and Objects.requireNonNull. This comment should be applied for every static string in this file in a class attribute.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/ExecutionComparisonView.java:
```diff
@@ -0,0 +1,1212 @@
+    private TmfTimestampFormat fFormat = new TmfTimestampFormat("yyyy-MM-dd HH:mm:ss.SSS.SSS.SSS"); //$NON-NLS-1$
+    private @Nullable Listener fSashDragListener;
+    private @Nullable SashForm fsashForm = null;
+    private static final String TMF_VIEW_UI_CONTEXT = "org.eclipse.tracecompass.tmf.ui.view.context"; //$NON-NLS-1$
```
All the static strings and all the static attributes should be grouped together.

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/ExecutionComparisonView.java:
```diff
@@ -0,0 +1,1212 @@
+    }
+
+    @Override
+    public void createPartControl(@Nullable Composite parent) {
```
All the comments delimit a subsection and should be refactored in a separate method to reduce the complexity.

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/ExecutionComparisonView.java:
```diff
@@ -0,0 +1,1212 @@
+
+    }
+
+    private void buildDifferetialFlameGraph() {
```
Correct typo

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/ExecutionComparisonView.java:
```diff
@@ -0,0 +1,1212 @@
+        }
+    }
+
+    private TmfXYChartViewer getChartViewerA() {
```
All of these private "accessors" should be deleted.

vladarama replied:
Removing these private accessors and using the class fields directly causes issues since these fields are `@Nullable`. Making them `@NonNull` also causes issues since we can't instantiate them in the constructor. 

What do you think is the best way of proceeding ? I can see two options but maybe I'm missing something:
1 - Keep the private accessors
2 - Use `Objects.requireNonNull()`

arfio replied:
All of the @Nullable should be kept where necessary. I do not believe most of those attributes are necessary as most of them are accessed in the createPartControl method only. The methods should all be deleted and, in most methods the values should be passed as parameters.

For example, the methods createResetButtonA, and B are mostly duplicated code which uses attributes whereas it should be using only parameters and this fixes the nullable issue because in createPartControl, we know that those values are not null because they are assigned before. This makes the static analysis more straight forward and avoid errors. 

Also in signal handlers, the fact that the values are nullable should be handled individually as right now, even though there are no warnings we could have some race conditions which would throw errors. Hiding them under functions and requireNonNull is not a good way of fixing those.

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/ExecutionComparisonView.java:
```diff
@@ -0,0 +1,1212 @@
+        fContextService = contextService;
+    }
+
+    private String makeQuery() {
```
The method's name should be changed to something more explicit and a comment may be used to explain what this method does.

vladarama replied:
Done 👍

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/ExecutionComparisonView.java:
```diff
@@ -0,0 +1,1212 @@
+        return query.toString();
+    }
+
+    boolean parseQuery(String query) {
```
Same comment as above

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1255 @@
+    /**
+     * ID of the view
+     */
+    public static final String ID = DifferentialFlameGraphView.class.getPackage().getName() + ".diffflamegraphView"; //$NON-NLS-1$
```
All of the class attributes should be grouped by category make it more readable. (every static final, every providers, every locks, ...

arfio commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.ui/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/ui/DifferentialFlameGraphView.java:
```diff
@@ -0,0 +1,1255 @@
+            if (trace == null) {
+                return;
+            }
+            // If entries for this trace are already available, just zoom on
```
This comment should be done using /* */

arfio reviewed (changes requested):
Lots and lots of small comments, most are easy to fix, others require more thinking/design. Also all of the files are contained under "analyses/org.eclipse.tracecompass.incubator.executioncomparision..." These folders should be renamed to correct the typo. You should remove and reimport the files when doing that change so that eclipse does not recreate these folders.
Good luck !

vladarama commented on analyses/org.eclipse.tracecompass.incubator.executioncomparision.core/src/org/eclipse/tracecompass/incubator/internal/executioncomparison/core/ParametricWeightedTreeUtils.java:
```diff
@@ -0,0 +1,289 @@
+ * @author Fateme Faraji Daneshgar
+ */
+@SuppressWarnings("restriction")
+public final class ParametricWeightedTreeUtils {
```
Done, removed `ParametricWeightedTreeUtils` and modified `WeightedTreeUtils` to achieve the same functionality.

vladarama commented:
@arfio 
Thanks for the review,  I finished addressing your comments. 
I left some questions in the `Conversation` tabs. I would appreciate a second review when you have the time.

arfio reviewed (dismissed):
Thank you for the overhaul ! This change is ready to be merged for me. Some refactors might still be needed in the future to improve code maintainability, specially for the view classes but believe it can be merged in that state.

vladarama commented:
@arfio 
Thanks for the review. I squashed my commits, can you reapprove the changes when you get the chance ?

bhufmann commented on analyses/org.eclipse.tracecompass.incubator.executioncomparison.ui.swtbot.tests/build.properties:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: EPL-2.0
+###############################################################################
+
+source.. = src/
```
My IDE shows an error because src directory doesn't exists

vladarama replied:
Removed `swtbot.tests` 👍

bhufmann reviewed (commented):
When building the incubator RCP, this feature is not included. I'm only able to try it using my development environment. Should it be added to the incubator RCP or are there follow-up patches planned for that?

It would be good to have some documentation on how to use it. This can be also part of a follow-up patch.

bhufmann commented on analyses/pom.xml:
```diff
@@ -39,6 +39,10 @@
+    <module>org.eclipse.tracecompass.incubator.executioncomparison</module>
+    <module>org.eclipse.tracecompass.incubator.executioncomparison.core</module>
+    <module>org.eclipse.tracecompass.incubator.executioncomparison.core.tests</module>
+    <module>org.eclipse.tracecompass.incubator.executioncomparison.ui</module>
```
Thanks for adding this in the last updated. With this the modules are built with maven.

bhufmann commented:
> @bhufmann @arfio I think we can merge this patch as is and do separate follow-up patches for the RCP, update site and documentation which would be easier to review than keeping everything in the same patch.

Make sense. I'm ok with this. Please include the plugins and feature to the analysis/pom.xml. If it builds successfully, then we can merged.
